// typed-table/lib/typed-table.js
// Generated by CoffeeScript 1.9.1
'use strict';
var BASE_DIR, CHAR_CODE_A, CHAR_CODE_Z, Cell, Jaco, NAME_COLUMN_VALUES, NAME_COLUMN_VALUES_LENGTH, PARENT, Range, Sheet, TypedTable, _colorCodeToNumber, _getColFormNumber, _getNumberOfCol, fs, jaco, path, xlsx;

fs = require('fs');

path = require('path');

xlsx = require('xlsx');

jaco = require('jaco');

Jaco = jaco.Jaco;

PARENT = module.parent;

BASE_DIR = path.dirname(PARENT.filename);

CHAR_CODE_A = 64;

CHAR_CODE_Z = 90;

NAME_COLUMN_VALUES = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split('');

NAME_COLUMN_VALUES_LENGTH = NAME_COLUMN_VALUES.length;

_getNumberOfCol = function(r1c1) {
  var c, i, k, len, n, num, s;
  num = 0;
  s = r1c1.toUpperCase().split('');
  for (i = k = 0, len = s.length; k < len; i = ++k) {
    c = s[i];
    n = NAME_COLUMN_VALUES.indexOf(c);
    if (1 !== 0 || i < s.length - 1) {
      n++;
    }
    num = num * 26 + n;
  }
  return num;
};

_getColFormNumber = function(num) {
  var col, mod, s;
  s = '';
  col = 0;
  if (1 !== 0 || col > 0) {
    num--;
  }
  mod = num % NAME_COLUMN_VALUES_LENGTH;
  s = NAME_COLUMN_VALUES[mod] + s;
  num = Math.floor(num / NAME_COLUMN_VALUES_LENGTH);
  col++;
  while (num > 0) {
    if (1 !== 0 || col > 0) {
      num--;
    }
    mod = num % NAME_COLUMN_VALUES_LENGTH;
    s = NAME_COLUMN_VALUES[mod] + s;
    num = Math.floor(num / NAME_COLUMN_VALUES_LENGTH);
    col++;
  }
  return s;
};

_colorCodeToNumber = function(code) {
  return parseInt(code.replace(/#/, ''), 16);
};

Range = (function() {
  Range.prototype.startCol = null;

  Range.prototype.startRow = null;

  Range.prototype.endCol = null;

  Range.prototype.endRow = null;

  Range.prototype.startNCol = 0;

  Range.prototype.startNRow = 0;

  Range.prototype.endNCol = 0;

  Range.prototype.endNRow = 0;

  function Range(ref) {
    var refSplit;
    this.ref = ref;
    refSplit = /^([a-z]+)([0-9]+):([a-z]+)([0-9]+)/ig.exec(this.ref);
    this.startCol = refSplit[1];
    this.startRow = refSplit[2];
    this.endCol = refSplit[3];
    this.endRow = refSplit[4];
    this.startNCol = _getNumberOfCol(refSplit[1]);
    this.startNRow = +refSplit[2];
    this.endNCol = _getNumberOfCol(refSplit[3]);
    this.endNRow = +refSplit[4];
  }

  Range.prototype.toString = function() {
    return "" + this.startCol + this.startRow + ":" + this.endCol + this.endRow;
  };

  return Range;

})();

Cell = (function() {
  Cell.prototype.value = null;

  Cell.prototype.type = null;

  Cell.prototype.format = '';

  Cell.prototype.color = 0x000000;

  Cell.prototype.bgColor = -1;

  function Cell(value1, type1, format, color, bgColor) {
    this.value = value1;
    this.type = type1;
    this.format = format;
    this.color = color != null ? color : 0x000000;
    this.bgColor = bgColor != null ? bgColor : -1;
  }

  return Cell;

})();

Sheet = (function() {
  Sheet.prototype.range = null;

  Sheet.prototype.rows = null;

  function Sheet(sheetData) {
    var c, cellData, cellValue, cl, cols, id, r, rl;
    this.rows = [];
    this.range = new Range(sheetData['!ref']);
    r = this.range.startNRow;
    rl = this.range.endNRow;
    while (r <= rl) {
      c = this.range.startNCol;
      cl = this.range.endNCol;
      cols = [];
      while (c <= cl) {
        id = "" + (_getColFormNumber(c)) + r;
        cellData = sheetData[id];
        if (cellData) {
          cellValue = new Cell(cellData.v, cellData.t, cellData.f);
          if (cellData.s) {
            cellValue.bgColor = parseInt(cellData.s.fgColor.rgb, 16);
          }
        } else {
          cellValue = null;
        }
        cols[c - 1] = cellValue;
        c++;
      }
      this.rows[r - 1] = cols;
      r++;
    }
  }

  return Sheet;

})();

TypedTable = (function() {
  TypedTable.prototype.rows = null;

  TypedTable.prototype.header = null;

  TypedTable.prototype.types = null;

  function TypedTable(rows, rowOption) {
    var DESC_ROW_NUM, HEADER_ROW_NUM, LABEL_ROW_NUM, TYPE_ROW_NUM, cols, k, len, rowNum;
    rowOption = rowOption || {};
    LABEL_ROW_NUM = rowOption.label !== void 0 ? rowOption.label : 0;
    HEADER_ROW_NUM = rowOption.header !== void 0 ? rowOption.header : 1;
    TYPE_ROW_NUM = rowOption.type !== void 0 ? rowOption.type : 2;
    DESC_ROW_NUM = rowOption.description !== void 0 ? rowOption.description : null;
    this.rows = [];
    this.header = [];
    this.types = [];
    for (rowNum = k = 0, len = rows.length; k < len; rowNum = ++k) {
      cols = rows[rowNum];
      switch (rowNum) {
        case LABEL_ROW_NUM:
        case DESC_ROW_NUM:
          continue;
        case HEADER_ROW_NUM:
          this.header = cols.slice(0);
          break;
        case TYPE_ROW_NUM:
          this.types = cols.slice(0);
          break;
        default:
          this.rows.push(cols.slice(0));
      }
      cols = null;
    }
    rows = null;
  }

  TypedTable.prototype.toJSON = function() {
    var allNull, arr, cell, cellValues, childName, data, headerName, i, item, j, k, keyName, keyNameSplit, l, len, len1, parentName, ref, row, type, value;
    data = [];
    allNull = true;
    ref = this.rows;
    for (i = k = 0, len = ref.length; k < len; i = ++k) {
      row = ref[i];
      cellValues = {};
      for (j = l = 0, len1 = row.length; l < len1; j = ++l) {
        cell = row[j];
        headerName = this.header[j].value;
        if (this.types && this.types[j]) {
          type = this.types[j].value;
        }
        if (headerName) {
          keyName = headerName.trim().replace(/\s/gm, ' ');
          keyName = new Jaco(keyName).toNarrow().toWideKatakana().toString();
          value = (function() {
            if (cell) {
              switch (String(type).toLowerCase()) {
                case 'c':
                case 'color':
                case 'colour':
                  if (/^#(?:[0-9a-f]{3}|[0-9a-f]{6})$/i.test(cell.value)) {
                    return _colorCodeToNumber(cell.value);
                  } else {
                    return cell.bgColor;
                  }
                  break;
                case 'd':
                case 'date':
                case 't':
                case 'time':
                  return new Date((+cell.value - 25569) * 86400 * 1000 || 0);
                case 'a':
                case 'arr':
                case 'ary':
                case 'array':
                  return arr = (function() {
                    var len2, m, ref1, results;
                    ref1 = ('' + cell.value).split(',');
                    results = [];
                    for (m = 0, len2 = ref1.length; m < len2; m++) {
                      item = ref1[m];
                      if (item !== '') {
                        results.push(item.trim());
                      }
                    }
                    return results;
                  })();
                case 'b':
                case 'bool':
                case 'boolean':
                  return !!cell.value;
                case 'n':
                case 'num':
                case 'number':
                  return +cell.value;
                default:
                  return ('' + cell.value).replace(/\r/, '');
              }
            } else {
              return null;
            }
          })();
          if (keyName.match(/^[a-z][a-z0-9_-]+--/ig)) {
            keyNameSplit = keyName.split(/--/);
            parentName = keyNameSplit[0] + 's';
            childName = keyNameSplit[1];
            if (cellValues[parentName] == null) {
              cellValues[parentName] = {};
            }
            cellValues[parentName][childName] = value;
          } else {
            cellValues[keyName] = value;
          }
          if (value !== null && allNull === true) {
            allNull = false;
          }
        }
      }
      if (!allNull) {
        data.push(cellValues);
      }
      allNull = true;
    }
    return data;
  };

  TypedTable.prototype.toJSONStringify = function(replacer, space) {
    if (space == null) {
      space = '	';
    }
    return JSON.stringify(this.toJSON(), replacer, space);
  };

  TypedTable.prototype.saveJSONFile = function(fileName, space) {
    var json;
    if (space == null) {
      space = '	';
    }
    json = this.toJSONStringify(null, '	');
    fs.writeFileSync(fileName, json);
  };

  TypedTable.readExcel = function(xlsxFile, rowOption) {
    var file, filePath, k, len, name, ref, sheet, sheets, table, tables;
    if (path.isAbsolute(xlsxFile)) {
      filePath = xlsxFile;
    } else {
      filePath = path.join(process.cwd(), xlsxFile);
    }
    file = xlsx.readFile(filePath, {
      cellStyles: true,
      cellNF: true
    });
    sheets = [];
    tables = [];
    ref = file.SheetNames;
    for (k = 0, len = ref.length; k < len; k++) {
      name = ref[k];
      sheet = new Sheet(file.Sheets[name]);
      table = new TypedTable(sheet.rows, rowOption);
      sheets.push(sheet);
      tables.push(table);
    }
    return tables;
  };

  return TypedTable;

})();

module.exports = TypedTable;
